design/authn_azure/authn_azure_solution_design.md

```diff
@@ -0,0 +1,280 @@
+
+1. The two first general validations include validate_authenticator_exists and validate_security. 
+    
+    1. validate_authenticator_exists ensures that the authenticator sent as part of the request URI is one that we support (authn-k8s, authn-azure, etc.).
+
+    1. validate_security validates that the Azure authenticator has been defined correctly in policy and the host is permitted to authenticate to the webservice.
+
+1. The next validation, validate_credentials calls the authenticator's valid? function. If so, Azure-specific authenticator validations will be triggered. 
+
+    1. validate_azure_token (see below for a more detailed explanation of the function).
+    
+    1. validate_application_identity, the Conjur identity host annotations for the Azure VM and the properties extracted from the "xms_mirid" field of the Azure token provided by the Azure VM are compared (see below for a more detailed explanation of the function).
+
+1. Once the Azure authenticator-specific validations have passed, the general authenticator validations can continue.
+    1. validate_origin checks the origin of the request to determine if the origin of the request is restricted.
+    
+    1. audit_success writes the event to the audit logs
+
+1. If all validations pass, an audit entry is written to audit logs and a Conjur access token will be returned to the Azure requesting resource. This Conjur access token will be used for all further communications between the Azure resource and Conjur.
+
+#### Azure authenticator class
+ 
+ ```ruby
+require 'command_class'
+
+module Authentication
+  module AuthnAzure
+
+    Err = Errors::Authentication::AuthnAzure
+    Authenticator = CommandClass.new(
+      dependencies: {
+       ...
+      },
+      inputs: [:authenticator_input]
+    ) do
+      
+	  # Azure authenticator checks
+      def call
+        validate_azure_token
+        validate_application_identity
+      end
+
+      private
+
+      ...
+    end
+ ```
+ 
+#### validate_azure_token
+
+The Azure token validation checks the signature and claims of the Azure token to ensure that the token was issued by Azure Active Directory. To do this we need to:
+1. Discover the Oauth 2.0 provider - in this case the Azure AD is the Oauth 2.0 provider. Its URl will be retrieved from the "provider-uri" variable defined in the authenticator policy
+1. Retrieve JWKs from the discovery endpoint - These keys are used for the token validation
+1. Decode and verify the token using the JWKs.
+
+Steps 1 & 2 may seem familiar as they are the exact steps that we perform in the OIDC Authenticator. This is because OIDC is Oauth 2.0 based. We will take advantage on this and use the 3rd party used in the OIDC Authenticator to perform steps 1 & 2 in our authenticator.
+Unfortunately, the 3rd party used in the OIDC Authenticator does not implement an access token decoding. Thus, to perform step 3, we will need another 3rd party - the "jwt" gem. 
```

Comment:
@orenbm 
I think we should further make the difference b/w OIDC id token and Azure access token, fleshing out that the OIDC 3rd party offers a decoding for an ID token and not access token. Therefore, we need to look elsewhere.

Reply:
@orenbm make sure to use the same cache capabilities we used for oidc implementation, for example: we used concurrency_limited_cache to limit the amount of threads that can access in parallel to Azure, and in case of timeout between conjur to Azure we do not suffocate the server

Reply:
thanks @nessiLahav. I didn't add it to the doc to leave the code simple but i'll add it.